### ISSUE TITLE
Port kernel submodule of QtCore

### ISSUE THREAD

**Issue by dmik** [bitwiseworks/qtbase-os2#23]
Things still missing here after porting qmake are `QElapsedTimer`, `QSharedMemory`, `QSystemSemaphore` and the `QAbstractEvenDispatcher` implementation for OS/2 (was called `QEventDispatcherPM` in Qt4). Note that `QEventDispatcherPM` is needed for #20 (for the `QThread` implementation).

This task is part of #19.

**Comment by dmik** [comment]
Note that although we're taking `QEventDispatcherPM` from Qt4 as a base for the Qt5 implementation (where it will be called `QEventDispatcherOS2`, see below), there are a few things that are new and need to be changed accordingly. Event dispatching is part of the core functionality and it already involves PM, so it will take some time to understand the new logic and implement it. That's what I'm doing now.

Note that many class and function names containing `PM` in Qt4 will be renamed to have `OS2` instead of `PM` (this mostly relates to internal classes). The reason is that in Qt5 they dropped windowing manager separation (i.e. there are no `Q_WS_PM` or `Q_WS_WIN` macros any more) and only left the OS separation (e.g. `Q_OS_OS2`, `Q_OS_WIN32`). This is a good thing because an OS implied a single windowing manager in most cases anyway.

**Comment by dmik** [comment]
One of things changed in Qt5 is timer processing. E.g. timers now have [types](http://doc.qt.io/qt-5/qt.html#TimerType-enum).

**Comment by dmik** [comment]
Note that for `QElapsedTimer` we will stick with the generic implementation for now which uses system time for that (see [here](http://doc.qt.io/qt-5/qelapsedtimer.html#ClockType-enum) for more info). This basically means that the timer is not monotonic meaning that the elapsed time value will be affected if the user changes the time zone or if the daylight saving time mode changes in between.  Later we will improve it, created #24 for that — our primary task for now is to get it all up and running, optimizations are for later.

**Comment by dmik** [comment]
BTW, there will be one redesign decision compared to the event dispatcher implementation in Qt4. In Qt4 there was only one auxiliary thread that would call select in a loop (to implement integration of `QSocketNotifier` with the PM message queue) for the whole process. Now there will be one aux thread per each event dispatcher.

An event dispatcher is created for each application thread that wants to process Qt events. Usually,  there is only one thread in each Qt application that does this: thread 1 (the main application thread). This thread is also responsible for handling GUI input and paint requests. However, Qt allows to create independent event processing on any Qt thread. This is rarely used in real applications so the above change will normally not increase the total number of threads. On the other hand, this new design simplifies internal data structures (and gives some speedup) as there is no need to associate a PM message queue with each socket notifier and each timer created on behalf of the respective thread.

It will be always possible to slightly modify the new code to add support for many event dispatchers in a single aux thread was it was in Qt4 but I doubt we'll ever need this.

**Comment by dmik** [comment]
Note two things though:
- In fact, *each* thread started via `QThread` creates an event dispatcher (and a PM message queue). This was the same in Qt4.
- In the new design an auxiliary thread will be created only as needed: i.e. as long as there is at least one socket notifier or timer. And when the last one of them is removed, the thread will be automatically terminated. So if the thread doesn't use any timers or socket notifiers, no auxiliary thread will be created at all, only a PM message queue.

**Comment by dmik** [comment]
BTW, some other important details of the new event dispatcher implementation (related to the commit above):
- We don't use an auxiliary PM object window any more and post directly to the thread's message queue (which should save a bit of PM resources).
- We use Unix code as much as possible (in particular, we use Unix timer machinery almost unchanged to implement all kinds of Qt timers). But see also #24.

**Comment by dmik** [comment]
So far, missing from this submodule are OS/2 parts of the following: `QSharedMemory`, `QSystemSemaphore`.

**Comment by dmik** [comment]
In fact, these two classes are those which were never ported to OS/2 within Qt4. These are the two Qt4 tickets for that: http://trac.netlabs.org/qt4/ticket/10 and http://trac.netlabs.org/qt4/ticket/11. `QSharedMemory` is rather simple, however, `QSystemSemaphore` is a bit of a work. I will re-consider doing it now to avoid breaking dependencies (examples, tests).

**Comment by dmik** [comment]
Actually, I came up with a nice and simple solution for `QSystemSemaphore` instead of what I proposed in http://trac.netlabs.org/qt4/ticket/11 (see above). I simply use a regular OS/2 event semaphore with some simple magic. This saved us a lot of work.

The idea is that OS/2 maintains a posted count associated with each semaphore (including the shared ones). I use this count to implement the number of resources needed for `QSystemSemaphore` logic. I didn't even need another mutex in order to implement this and it should be fully thread-safe and process-wide as well. The only downside of this approach is a possible deadlock situation if a process unexpectedly crash, but it's assumed by the Qt API to some degree, see #27 for more info. Anyway, for now it's more than fine.

And `QSharedMemory` was easy once we got `QSystemSemaphore`.
